src/clib/initialize_chemistry_data.c

```diff
@@ -238,13 +238,63 @@ int local_initialize_chemistry_data(chemistry_data *my_chemistry,
     return FAIL;
   }
 
-  if (my_chemistry->primordial_chemistry == 0) {
+  // deal with my_chemistry->HydrogenFractionByMass
+  // ==============================================
+  // - set_default_chemistry_parameters & local_initialize_chemistry_parameters
+  //   historically configured my_chemistry such that this member had a value
+  //   of 0.76 (the default when primordial_chemistry >= 1). If
+  //   primordial_chemistry == 0, this would always silently overwrite the
+  //   value in fully tabulated mode. This led to cases where downstream users
+  //   might mistakenly think they modified this parameter
+  //
+  // - New behavior: the member is now initialized to an undefined value
+  //   - if the user hasn't changed it, we now set it to the appropriate
+  //     default value
+  //   - if the user has changed it, the precise handling depends on the choice
+  //     of primordial_chemistry
+
+  if (my_chemistry->HydrogenFractionByMass > 1) {
+    fprintf(stderr, "ERROR: HydrogenFractionByMass cannot exceed 1.0\n");
+    return FAIL;
+  } else if (my_chemistry->primordial_chemistry == 0) {
     /* In fully tabulated mode, set H mass fraction according to
        the abundances in Cloudy, which assumes n_He / n_H = 0.1.
-       This gives a value of about 0.716. Using the default value
+       This gives a value of about 0.716. Using the other default value
        of 0.76 will result in negative electron densities at low
-       temperature. Below, we set X = 1 / (1 + m_He * n_He / n_H). */
-    my_chemistry->HydrogenFractionByMass = 1. / (1. + 0.1 * 3.971);
+       temperature. Below, use X = 1 / (1 + m_He * n_He / n_H). */
+
+    // we precomputed this value (in enhanced precision) rather than compute
+    // it on the fly to allow users to directly specify this value (without
+    // worrying about round-off issues)
+    const double default_Hfrac = 0.715768377353088514;
+    // TODO: at some point in the future, we should store the above value as
+    //       an attribute of the HDF5 table and read it in directly.
+
+    if (my_chemistry->HydrogenFractionByMass < 0) {
+      my_chemistry->HydrogenFractionByMass = default_Hfrac;
+    } else if (my_chemistry->HydrogenFractionByMass != default_Hfrac) {
+      fprintf(stderr,
+              "ERROR: Invalid HydrogenFractionByMass value is specified.\n"
+              " -> when primordial_chemistry == 0, the allowed values are\n"
+              "    are strictly enforced.\n"
+              " -> the specified value is: %.18f\n"
+              " -> it is allowed to have a negative value. In this case, the\n"
+              "    value is overwritten to match the tabulated default\n"
+              "    value of %.18f that is consistent with assumptions made\n"
+              "    while constructing the table (this is the default \n"
+              "    behavior if you don't touch the value of this member.)\n"
+              " -> alternatively you can EXACTLY specify the default value\n"
```

Comment:
I am generally in favor of this change, but I find the wording here a little hard to parse. Could this be reworded to state the available options a bit more plainly? For me, the main things are:
- leave the value unset or set to negative number to have it calculated internally as ###
- set the value exactly to the above number

Reply:
Thanks for the feedback. I think I addressed your concerns